cmd/private-prow-configs-mirror/main.go

```diff
@@ -0,0 +1,437 @@
+	}
+}
+
+func injectPrivatePlugins(plugins map[string][]string, orgRepos orgReposWithOfficialImages) {
+	logrus.Info("Processing...")
+
+	var allPrivate []string
+
+	privateReposWithValues := make(map[string][]string)
+
+	for orgRepo, value := range plugins {
+		if orgRepos.isOfficialRepoFull(orgRepo) {
+			newValue := value
+
+			// We want to append the org level values, because the private repo
+			// can exist in different org with different configuration.
+			org := strings.Split(orgRepo, "/")[0]
+			if _, ok := plugins[org]; ok {
+				newValue = append(newValue, plugins[org]...)
+			}
+
+			privateReposWithValues[privateOrgRepo(strings.Split(orgRepo, "/")[1])] = newValue
+			allPrivate = append(allPrivate, newValue...)
+		}
+	}
+
+	mergeCommonPrivatePlugins(plugins, privateReposWithValues, getRepeatedValues(allPrivate))
+}
+
+// mergeCommonPrivatePlugins detects the common values from all the private repositories
+// and merges them into the org `openshift-priv` level.
+// In addition, it generates a repo. if it contains an extra non-common field.
+func mergeCommonPrivatePlugins(plugins map[string][]string, privateReposWithValues map[string][]string, repeatedValues sets.String) {
+	privateOrgPlugins := sets.NewString()
+
+	for repoName, repoValues := range privateReposWithValues {
+		repoLevelPlugins := sets.NewString()
+		valuesSet := sets.NewString(repoValues...)
+
+		if len(repeatedValues) > 0 {
+			commonValues := valuesSet.Intersection(repeatedValues)
+			privateOrgPlugins.Insert(commonValues.List()...)
+
+			if diffValues := valuesSet.Difference(repeatedValues); len(diffValues) > 0 {
+				repoLevelPlugins.Insert(diffValues.List()...)
+			}
+
+		} else {
+			repoLevelPlugins.Insert(repoValues...)
+		}
+
+		if len(repoLevelPlugins.List()) > 0 {
+			logrus.WithFields(logrus.Fields{"repo": repoName, "value": repoLevelPlugins.List()}).Info("Generating repo")
+			plugins[repoName] = repoLevelPlugins.List()
+		}
+	}
+
+	if len(privateOrgPlugins.List()) > 0 {
+		logrus.WithField("value", privateOrgPlugins.List()).Info("Generating openshift-priv org.")
+		plugins[openshiftPrivOrg] = privateOrgPlugins.List()
+	}
+}
+
+func getRepeatedValues(values []string) sets.String {
+	temp := sets.NewString()
+	repeated := sets.NewString()
+
+	for _, value := range values {
+		if temp.Has(value) {
+			repeated.Insert(value)
+		} else {
+			temp.Insert(value)
+		}
+	}
+
+	return repeated
+}
+
+func main() {
+	logrus.SetReportCaller(true)
+	logrus.SetFormatter(&logrus.TextFormatter{
+		CallerPrettyfier: func(f *runtime.Frame) (string, string) {
+			return fmt.Sprintf("%s()", f.Function), fmt.Sprintf("%s:%d", path.Base(f.File), f.Line)
+		},
+	})
+
+	o := gatherOptions()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid option")
+	}
+
+	go func() {
+		interrupts.WaitForGracefulShutdown()
+		os.Exit(1)
+	}()
+
+	currentDir, err := os.Getwd()
```

Comment:
This does not match the expectations given by options of the program... I tried to test and I got a crash:

```console
$ go run ./cmd/private-prow-configs-mirror/ --ci-op-config-path ../release/ci-operator/config/ --prow-config=../release/core-services/prow/02_config/_config.yaml --prow-plugins ../release/core-services/prow/02_config/_plugins.yaml
ERRO[0000]load.go:159 github.com/openshift/ci-tools/pkg/config.OperateOnCIOperatorConfigSubdir.func1() Failed to walk CI Operator configuration dir  error="lstat /home/remote/pmuller/Projects/RH/go/src/github.com/openshift/ci-tools/ci-operator/config: no such file or directory" source-file=/home/remote/pmuller/Projects/RH/go/src/github.com/openshift/ci-tools/ci-operator/config
WARN[0000] failed to load ci-operator configuration from release repo  error="lstat /home/remote/pmuller/Projects/RH/go/src/github.com/openshift/ci-tools/ci-operator/config: no such file or directory"
WARN[0000] failed to load Prow configuration from release repo  error="stat /home/remote/pmuller/Projects/RH/go/src/github.com/openshift/ci-tools/core-services/prow/02_config/_config.yaml: no such file or directory"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13ebd94]

goroutine 1 [running]:
main.main()
	/home/remote/pmuller/Projects/RH/go/src/github.com/openshift/ci-tools/cmd/private-prow-configs-mirror/main.go:410 +0x484
exit status 2
```

Don't assume to run in the correct cwd... The inputs and ouputs need to be more systematic (and documented). At the very least the path to relesae repo needs to be a param. We need full paths to the configs anyway (because we need to write to them), so I think we should just read the configs directly ourselves and not use `GetAllConfigs()` helper.